src/test/groovy/org/prebid/server/functional/tests/pg/TargetingFirstPartyDataSpec.groovy

```diff
@@ -0,0 +1,401 @@
+        given: "Bid request"
+        def bidRequest = BidRequest.defaultBidRequest.tap {
+            imp = [Imp.defaultImpression.tap {
+                banner = Banner.defaultBanner
+                ext.context = new ImpExtContext(data: impExtContextData)
+            }]
+        }
+
+        and: "Planner response"
+        def plansResponse = PlansResponse.getDefaultPlansResponse(bidRequest.site.publisher.id).tap {
+            lineItems[0].targeting = Targeting.defaultTargetingBuilder
+                                              .addTargeting(targetingType, INTERSECTS, [integerTargetingValue])
+                                              .build()
+        }
+        generalPlanner.initPlansResponse(plansResponse)
+
+        and: "Line items are fetched by PBS"
+        updateLineItemsAndWait()
+
+        when: "Auction is happened"
+        def auctionResponse = pgPbsService.sendAuctionRequest(bidRequest)
+
+        then: "PBS had PG auction"
+        assert auctionResponse.ext?.debug?.pgmetrics?.matchedWholeTargeting?.size() == plansResponse.lineItems.size()
+
+        where:
+        targetingType  | impExtContextData
+        SFPD_BUYER_ID  | new ImpExtContextData(buyerid: integerTargetingValue)
+        SFPD_BUYER_IDS | new ImpExtContextData(buyerids: [integerTargetingValue])
+    }
+
+    def "PBS shouldn't throw a NPE for '#targetingType' when its Ext is absent and targeting Intersects matching type is selected"() {
+        given: "Planner response"
+        def plansResponse = PlansResponse.getDefaultPlansResponse(bidRequest.site.publisher.id).tap {
+            lineItems[0].targeting = Targeting.defaultTargetingBuilder
+                                              .addTargeting(targetingType, INTERSECTS, [stringTargetingValue])
+                                              .build()
+        }
+        generalPlanner.initPlansResponse(plansResponse)
+
+        and: "Line items are fetched by PBS"
+        updateLineItemsAndWait()
+
+        when: "Auction is happened"
+        def auctionResponse = pgPbsService.sendAuctionRequest(bidRequest)
+
+        then: "PBS successfully processed request"
+        notThrown(PrebidServerException)
+
+        and: "PBS hasn't had PG auction as request targeting is not specified in the right place"
+        assert !auctionResponse.ext?.debug?.pgmetrics
+
+        where:
+        targetingType | bidRequest
+        SFPD_KEYWORDS | BidRequest.defaultBidRequest.tap {
+            site = Site.defaultSite.tap {
+                keywords = stringTargetingValue
+            }
+        }
+
+        UFPD_LANGUAGE | BidRequest.defaultBidRequest.tap {
+            user = User.defaultUser.tap {
+                language = stringTargetingValue
+            }
+        }
+    }
+
+    def "PBS should be able to match site FPD targeting taken from different sources by INTERSECTS matching function"() {
```

Comment:
Newly added tests to cover alt input sources to match PG targeting starts from that test in that spec.